### PR TITLE
Tools: Topology: Add ASRC capture to cavs-nocodec.conf topologies

### PR DESCRIPTION
**Pull request by andrula-song**
This patch adds ASRC capture from Dmic such as sof-tgl-nocodec.tplg. Set ASRC_CAPTURE_ENABLE as true enables test of ASRC component with IPC4.

### CODE REVIEW

**Comment by andrula-song** [comment]
add pipeline 23, capture pipeline from dmic0-> asrc->host-copier
sof-tgl-nocodec

File: tools/topology/topology2/cavs-nocodec.conf
Comment (lgirdwood) [inline tools/topology/topology2/cavs-nocodec.conf:361]: indentation ?

**Review by lgirdwood — COMMENTED** [review]
@singalsu @jsarha @ranj063 pls review

File: tools/topology/topology2/cavs-nocodec.conf
Comment (jsarha) [reply tools/topology/topology2/cavs-nocodec.conf:361]: Do I read this correctly, we have?
```
IncludeByKey.PASSTHROUGH {
"false" {
	...
	IncludeByKey.NUM_DMICS {
	"[1-4]"	{
		...
		}
	}

+	IncludeByKey.NUM_DMICS {
+	"[1-4]" {
+		IncludeByKey.ASRC_CAPTURE_ENABLE {
+		"true"	{
+			...
+			}
+		}
+	}
	...
```

Is there some reason for having two NUM_DMIC [1-4] blocks following each other, or is this use of big nested IncludebyKey blocks just getting out of hand?

Not in the scope of this PR, but it could be a time to split nocodec file into smaller pieces included from within IncludeByKey blocks.

**Comment by plbossart** [comment]
I still have concerns about 
a) why we think ASRC is a good idea for a nocodec topology where all clocks are derived from the same master clock...
b) how ASRC is actually tested and what the value of this integration is.

NAK for now.

File: tools/topology/topology2/include/components/asrc_format_s32_convert_from_24k.conf
Comment (singalsu) [inline tools/topology/topology2/include/components/asrc_format_s32_convert_from_24k.conf:1]: Are these the same as with SRC, could the same .conf files be used for both?

**Comment by singalsu** [comment]
> I still have concerns about a) why we think ASRC is a good idea for a nocodec topology where all clocks are derived from the same master clock... b) how ASRC is actually tested and what the value of this integration is.
> 
> NAK for now.

Could the topology for nocodec be like this:

```
host copier <--- ASRC <--- DAI copier DMIC

host copier ---> ASRC ---> DAI copier SSP0
host copier <------------- DAI copier SSP0

host copier -------------> DAI copier SSP2
host copier <--- ASRC <--- DAI copier SSP2

host copier ---> ASRC ---> DAI copier SSP1
host copier <--- ASRC <--- DAI copier SSP1
```
It would allow to capture back playback ASRC output in SSP loopback mode (LBM). And feed data for capture ASRC.

The SSP1 duplex ASRC is to test shared DAI tracking. SSP1 is not working in UPX so some other DUT needs to be used or switch SSPs to test such after the two simpler cases are tested to work.

The default SSP configuration in nocodec is clock producer while we would need follower role. But this would be a start until we figure out how to get follower mode to work and how to hook a clock producer sound card to UPX audio HAT.

I tested ASRC originally with UP2 and Digiberry S/PDIF sound card and Pierre's SSP follower setup. I used S/PDIF connection to a USB sound card. It drifted in rate slightly vs. timer scheduled SOF.

It would be a fair effort to try to achieve similar setup with UPX. Do you think we should do it or just stick in tracking a synchronous DAI and believing that it would track a clock follower DAI?

**Comment by andrula-song** [comment]
the latest topology:
sof-tgl-nocodec-asrc

File: tools/topology/topology2/include/pipelines/cavs/dai-copier-asrc-capture.conf
Comment (aiChaoSONG) [inline tools/topology/topology2/include/pipelines/cavs/dai-copier-asrc-capture.conf:52]: Those two attributes could be transformed to string type to make it human readable.

File: tools/topology/topology2/include/pipelines/cavs/dai-copier-asrc-capture.conf
Comment (aiChaoSONG) [inline tools/topology/topology2/include/pipelines/cavs/dai-copier-asrc-capture.conf:27]: duplicated definition with pipeline-common.conf

File: tools/topology/topology2/development/cavs-nocodec-asrc.conf
Comment (singalsu) [inline tools/topology/topology2/development/cavs-nocodec-asrc.conf:264]: Is this a capture ASRC with 44.1 kHz rate out to PCM?

File: tools/topology/topology2/development/cavs-nocodec-asrc.conf
Comment (singalsu) [inline tools/topology/topology2/development/cavs-nocodec-asrc.conf:541]: Is they playback DAI really 44.1 kHz?

File: tools/topology/topology2/include/components/asrc_format_s32_to_sxx_convert.conf
Comment (singalsu) [inline tools/topology/topology2/include/components/asrc_format_s32_to_sxx_convert.conf:6]: These could use the new array definition in alsatplg. Can these be shared, same include files, with SRC?

File: tools/topology/topology2/include/pipelines/cavs/asrc-dai-copier-playback.conf
Comment (singalsu) [inline tools/topology/topology2/include/pipelines/cavs/asrc-dai-copier-playback.conf:86]: Why 3 input audio formats and one output? The asrc is not supporting format conversion, but same input and output format for s16/s24/s32.

File: tools/topology/topology2/development/cavs-nocodec-asrc.conf
Comment (andrula-song) [reply tools/topology/topology2/development/cavs-nocodec-asrc.conf:541]: sorry, used don't know that SSP can not do sample convert.

File: tools/topology/topology2/development/cavs-nocodec-asrc.conf
Comment (andrula-song) [reply tools/topology/topology2/development/cavs-nocodec-asrc.conf]: No, it is 48kHz

File: tools/topology/topology2/development/cavs-nocodec-asrc.conf
Comment (singalsu) [inline tools/topology/topology2/development/cavs-nocodec-asrc.conf:531]: In my UPX SSP0 and SSP2 work, SSP1 doesn't, I have used used SSP0 and SSP2 in topologies for tgl  but I guess your test device is different.

**Review by singalsu — APPROVED** [review]
Looks good to me

File: tools/topology/topology2/development/cavs-nocodec-asrc.conf
Comment (andrula-song) [reply tools/topology/topology2/development/cavs-nocodec-asrc.conf:531]: yes, I can get the right curve

**Comment by andrula-song** [comment]
hello @ranj063 , can you help to review this PR?

**Comment by lgirdwood** [comment]
@andrula-song conflicts.

**Comment by lgirdwood** [comment]
@andrula-song @singalsu @plbossart my preference is to maximise code under regular test, even if we have to fake the remote time domain here in this mode.

Thinking aloud we could have a test Kconfig for ASRC that fakes the remote time domain. i.e. we can read CCOUNT (via Zephyr API) and convert some of it's LSBs to a random drift value.

File: tools/topology/topology2/include/components/asrc.conf
Comment (kv2019i) [inline tools/topology/topology2/include/components/asrc.conf:122]: This is a bit odd, why does the UUID change?

File: tools/topology/topology2/include/components/asrc_format_s32_convert_from_24k.conf
Comment (kv2019i) [inline tools/topology/topology2/include/components/asrc_format_s32_convert_from_24k.conf:38]: Ditto here, 1 format, but below there are 3 formats listed.

**Review by kv2019i — CHANGES_REQUESTED** [review]
Few minor notes on the tplg code

**Comment by plbossart** [comment]
> @andrula-song @singalsu @plbossart my preference is to maximise code under regular test, even if we have to fake the remote time domain here in this mode.

Of course we want to test as many things as possible, but we are both resource-constrained and time-constrained. 

The fundamental issue is IMHO a confusion between component testing and CI (Continuous Integration). We have exactly zero products that rely on ASRC, so while it's an important capability it has by construction a lower priority. I will assert that ASRC does not belong in the nocodec topology that is the basic for our quick smoke tests for hardware (DMA, DAIs, etc). Exhibit A for this assertion is what we do for each power-on: we fall-back to the traditional passthrough pipelines and gradually add complexity.

And even if this ASRC component is added as an opt-in, it's not clear to me if/when this component would be tested, by what/whom and at which frequency.

I would like the test methodology to be addressed before we merge anything. "Merge first, think later about test" will lead to bitrot or be dependent on developer bandwidth, both of which are not desirable directions.

**Comment by singalsu** [comment]
@plbossart The name of this PR is somewhat confusing, the usual nocodec topology we use is not impacted, this adds another nocodec type topology to use alternatively. It can be seen from the CMakeLists.txt part or PR `development/cavs-nocodec-asrc\;sof-tgl-nocodec-asrc`. The source .conf is different and the produced topology name is different.

**Comment by singalsu** [comment]
Even if this would not be used in regular test there should be a way to store a topology as evidence that the modifications to code (this time ipc4 & module adapter client) at least with that git hash worked. It's a lot of effort to create a new topology from scratch, and these topologies for test are very helpful when making the real production topologies. All the small details in getting things to work are quickly forgotten.

**Comment by plbossart** [comment]
@singalsu I can only concur with the notion that writing a new topology is hard, but so is maintaining a topology that has a gazillion options. It took us so much time to restore basic passthrough functionality which was broken by gradual additions. @ranj063 is still working on the format stuff, and we'll have even more changes coming.

I don't really follow the point of using a hash as a proof of the last working setup. We really need these files to be maintained, and that means agreeing on what gets tested and when.

File: tools/topology/topology2/include/components/asrc.conf
Comment (andrula-song) [reply tools/topology/topology2/include/components/asrc.conf:122]: Done, thanks!

**Comment by singalsu** [comment]
> It took us so much time to restore basic passthrough functionality which was broken by gradual additions.

It could be solved by splitting the pass-through topology into a separate file and build, like this PR. I've done it myself to understand what is needed for a bare bones topology and made some new topologies from the minimum. The normal cavs-nocodec.conf has grown beyond my comprehension. Changing it a bit too easily breaks something, at least when I'm trying.

**Comment by kv2019i** [comment]
@plbossart I think this matches e.g. situation with "sof-nocodec-bt-*.tplg". I'm in practise the only one who runs tests on this and tests run purely on a per-need basis. 

I'd still this is more useful than not having any topology at all. E.g. this is a reference of how to use a particular component in end-to-end context and with a topology than can be run on widely available hardware. Plus git will give you some indication on who to contact if you want run in to issue.

Would it help if we mandate a dedicated owner for each such topology? We could simply use CODEOWNERS to track owner.

**Review by kv2019i — APPROVED** [review]
Thanks @andrula-song , my concerns addressed.

**Comment by lgirdwood** [comment]
@plbossart good now ?

**Comment by plbossart** [comment]
@lgirdwood I am still opposed to the directions. This adds a new tools/topology/topology2/development/cavs-nocodec-asrc.conf file, and that brings up again the question on who is going to maintain yet another file.

I started talking with @ranj063 on a different approach where the host- and dai- copiers are included and only a simpler component added. This should be a pre-canned set of classes with a "PUT YOUR CODE HERE" when adding a new component, be it ASRC, DRC, and whatever is under development.

There is no reason to e.g. create yet another PCM device or replicate a complicated topology. We are again confusing component level testing and integration tests.

**Comment by lgirdwood** [comment]
> @lgirdwood I am still opposed to the directions. This adds a new tools/topology/topology2/development/cavs-nocodec-asrc.conf file, and that brings up again the question on who is going to maintain yet another file.
> 
> I started talking with @ranj063 on a different approach where the host- and dai- copiers are included and only a simpler component added. This should be a pre-canned set of classes with a "PUT YOUR CODE HERE" when adding a new component, be it ASRC, DRC, and whatever is under development.
> 
> There is no reason to e.g. create yet another PCM device or replicate a complicated topology. We are again confusing component level testing and integration tests.

Ok, I'm fine as long as we can have upstream topologies that can easily enable developers

1) Usage of any module under testbench with IPC4.
2) Usage of direct A->B IPC4 pipelines under plugin.

**Comment by singalsu** [comment]
@andrula-song There's a conflict

**Comment by andrula-song** [comment]
> @andrula-song There's a conflict

thanks, but since I have added the benchmark ASRC topologies, I would like to close this one.

**Pull request closed without merging**
